homeassistant/components/tasmota/fan.py

```diff
@@ -56,42 +59,40 @@ def __init__(self, **kwds):
         )
 
     @property
-    def speed(self):
-        """Return the current speed."""
-        return TASMOTA_TO_HA_SPEED_MAP.get(self._state)
+    def speed_count(self) -> Optional[int]:
+        """Return the number of speeds the fan supports."""
+        return len(ORDERED_NAMED_FAN_SPEEDS)
 
     @property
-    def speed_list(self):
-        """Get the list of available speeds."""
-        return list(HA_TO_TASMOTA_SPEED_MAP)
+    def percentage(self):
+        """Return the current speed percentage."""
+        if self._state is None:
+            return None
+        if self._state == 0:
+            return 0
+        return ordered_list_item_to_percentage(ORDERED_NAMED_FAN_SPEEDS, self._state)
 
     @property
     def supported_features(self):
         """Flag supported features."""
         return fan.SUPPORT_SET_SPEED
 
-    async def async_set_speed(self, speed):
+    async def async_set_percentage(self, percentage):
         """Set the speed of the fan."""
-        if speed not in HA_TO_TASMOTA_SPEED_MAP:
-            raise ValueError(f"Unsupported speed {speed}")
-        if speed == fan.SPEED_OFF:
+        if percentage == 0:
             await self.async_turn_off()
         else:
-            self._tasmota_entity.set_speed(HA_TO_TASMOTA_SPEED_MAP[speed])
-
-    #
-    # The fan entity model has changed to use percentages and preset_modes
-    # instead of speeds.
-    #
-    # Please review
-    # https://developers.home-assistant.io/docs/core/entity/fan/
-    #
+            tasmota_speed = percentage_to_ordered_list_item(
+                ORDERED_NAMED_FAN_SPEEDS, percentage
+            )
+            self._tasmota_entity.set_speed(tasmota_speed)
+
     async def async_turn_on(
         self, speed=None, percentage=None, preset_mode=None, **kwargs
     ):
         """Turn the fan on."""
         # Tasmota does not support turning a fan on with implicit speed
-        await self.async_set_speed(speed or fan.SPEED_MEDIUM)
+        await self.async_set_percentage(percentage or 66)
```

Comment:
Maybe this is more readable?
```py
await self.async_set_percentage(percentage or ordered_list_item_to_percentage(ORDERED_NAMED_FAN_SPEEDS, tasmota_const.FAN_SPEED_MEDIUM)
```

Reply:
That definitely shows the intent a bit better